### PR TITLE
Add Non Conservative Periodic Waveform Source for Twin Builder [SINE,…

### PR DESCRIPTION
**Pull request by SushobhitGupt**
… PULSE, TRIANG, SAWTOOTH]

### CODE REVIEW

File: pyaedt/modeler/PrimitivesTwinBuilder.py [hunk 1]
```diff
@@ -356,3 +356,73 @@ def create_pnp(self, compname=None, location=[], angle=0, use_instance_id_netlis
...
+        location=[],
+        angle=0,
+        use_instance_id_netlist=False,
+    ):
+        """
+        Create a periodic waveform source (non conservative real output).
+
+        Parameters
+        ----------
+        compname : str, optional
```
Comment (MaxJPRey) [inline pyaedt/modeler/PrimitivesTwinBuilder.py:379]: @SushobhitGupt Should we rename this argument ``source_name`` as it is the name of the voltage source. I understand that we are creating a component behind the scene. It is just to be sure.

**Review by MaxJPRey — APPROVED** [review]

File: pyaedt/modeler/PrimitivesTwinBuilder.py (hunk 1 above)
Comment (SushobhitGupt) [reply pyaedt/modeler/PrimitivesTwinBuilder.py:379]: @MaxJPRey compname is a known Twin Builder terminology, we probably would not want to change args based on what type of component is being placed (like sources in this case). keeping it same would keep consistency with other components defined in the file

Comment (MaxJPRey) [reply pyaedt/modeler/PrimitivesTwinBuilder.py:379]: @SushobhitGupt It makes sense.

File: pyaedt/modeler/PrimitivesTwinBuilder.py [hunk 2]
```diff
@@ -356,3 +356,73 @@ def create_pnp(self, compname=None, location=[], angle=0, use_instance_id_netlis
...
+        id = self.create_component(
+            compname,
+            component_library="Basic Elements\\Tools\\Time Functions",
+            component_name=type,
+            location=location,
+            angle=angle,
+            use_instance_id_netlist=use_instance_id_netlist,
+        )
+
+        if type == "SINE" or type == "PULSE" or type == "TRIANG" or type == "SAWTOOTH":
```
Comment (akaszynski) [inline pyaedt/modeler/PrimitivesTwinBuilder.py:419]:
```suggestion
        if type in ["SINE", "PULSE", "TRIANG", "SAWTOOTH"]:
```

**Review by maxcapodi78 — APPROVED** [review]
@SushobhitGupt maybe you could have a look also to components_catalog object that is now in schematic and which provides you the full list of components in Twinbuilder syslib and as well the possibility to place it. Anyway I think that most common used components should have dedicated method. I like it

**Comment by SushobhitGupt** [comment]
> @SushobhitGupt maybe you could have a look also to components_catalog object that is now in schematic and which provides you the full list of components in Twinbuilder syslib and as well the possibility to place it. Anyway I think that most common used components should have dedicated method. I like it

Thanks, I will check it out. As you mentioned, some commonly used components can be exposed so that we can create some good examples and tutorials. That is what my intent is here.

**Pull request merged**